### PR TITLE
Notification settings and OneSignal integration

### PR DESCRIPTION
**Pull request by matthewdias**
todo:
- [x] merge https://github.com/hummingbird-me/hummingbird-server/pull/200
- [x] set app id in env
- [x] complete steps [1 and 4](https://documentation.onesignal.com/docs/web-push-sdk-setup-https) to configure OneSignal
- [x] enable slideover
- [x] prompt after finishing feed onboarding 
- [x] customize messages
- [x] retries for player creation

things to consider:
- Once a user has disabled notifications after subscribing, there is apparently no way to resubscribe until they clear site data, even if they change permissions to allow them again. OneSignal treats the user as having blocked notifications, and doesn't attempt to resubscribe.
- We have the option of displaying a slidedown popup with a custom message before triggering the permission modal. The only problem with this is it will show if the user has blocked and reallowed notifications, so clicking confirm on the slidedown will actually fail to trigger the permission modal for the reason stated in the first bullet point.
- How do we handle a failed OneSignal player creation after enabling the permission?
- We might want to have another step in onboarding with a message about notifications and a button to trigger the permission modal.

### CODE REVIEW

File: app/controllers/application.js [hunk 1]
```diff
@@ -16,6 +16,24 @@ export default Controller.extend(queryParams.Mixin, {
   ajax: service(),
   store: service(),
 
+  init() {
+    this._super(...arguments);
+    // create the one-signal-player when user subscribes
+    window.OneSignal.push(() => {
+      window.OneSignal.on('subscriptionChange', (isSubscribed) => {
```
Comment (matthewdias) [inline app/controllers/application.js]: Is there a better place to put this? @vevix

**Comment by JoshFabian** [comment]
> We have the option of displaying a slidedown popup with a custom message before triggering the permission modal. The only problem with this is it will show if the user has blocked and reallowed notifications, so clicking confirm on the slidedown will actually fail to trigger the permission modal for the reason stated in the first bullet point.

Yeah, we definitely want to do this. It’s called “Permission priming”

> We might want to have another step in onboarding with a message about notifications and a button to trigger the permission modal.

It’ll actually be better to only show this to users when they hit a certain amount of engagement. I.E. they’ve finished onboarding / received their first notification.

> Once a user has disabled notifications after subscribing, there is apparently no way to resubscribe until they clear site data, even if they change permissions to allow them again. OneSignal treats the user as having blocked notifications, and doesn’t attempt to resubscribe.

For now, fuck it. We can come back to this one later.

> How do we handle a failed OneSignal player creation after enabling the permission?
Same deal, fuck it. We can come back to this after the big deadline has passed. Let’s get this up and running for the 99% first and fix the fringe stuff later.

File: app/controllers/application.js (hunk 1 above)
Comment (trmcnvn) [reply app/controllers/application.js]: Move to a private function in the route and call it from the authentication hook there.

May need to check that the application route isn't destroyed (I don't think it is, but don't know off top of my head)

**Comment by matthewdias** [comment]
ready for review @vevix

File: app/controllers/settings/notifications.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+      settingsTasks.push(get(this, 'updateSettingTask').perform(setting));
+    });
+    yield all(settingsTasks).then((results) => {
+      if (isEmpty(results.filter(result => result === undefined))) {
+        get(this, 'notify').success('Your notification settings were updated.');
+      }
+    });
+  }).drop(),
+
+  updateSettingTask: task(function* updateNotificationSetting(setting) {
```
Comment (trmcnvn) [inline app/controllers/settings/notifications.js]: Any reason why this function has a name?

File: app/storages/onesignal-setup.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+
+const Storage = StorageObject.extend();
+
+Storage.reopenClass({
+  initialState() {
+    return {};
+  }
+});
+
+export default Storage;
```
Comment (trmcnvn) [inline app/storages/onesignal-setup.js]: :scissors:

File: app/routes/application.js [hunk 4]
```diff
@@ -13,6 +13,7 @@ export default Route.extend(ApplicationRouteMixin, {
   metrics: service(),
   moment: service(),
   cache: storageFor('last-used'),
+  playerCache: storageFor('onesignal-setup'),
```
Comment (trmcnvn) [inline app/routes/application.js]: I think we have a `local-cache` service that we can use for this.

File: app/controllers/settings/notifications.js (hunk 2 above)
Comment (matthewdias) [reply app/controllers/settings/notifications.js]: whoops, bad refactor

**Comment by trmcnvn** [comment]
@matthewdias This good to be merged?

**Comment by matthewdias** [comment]
@vevix running into an issue where `public/OneSignalSDKUpdaterWorker.js` and `public/OneSignalSDKWorker.js` aren't being served on staging.

**Comment by trmcnvn** [comment]
They'll be getting hashed, you'll need to add them to the fingerprint blacklist under `ember-cli-build.js`

**Comment by trmcnvn** [comment]
They're being served now, had to clear CloudFront cache for them. Everything else :+1:?

**Comment by matthewdias** [comment]
👍

**Pull request merged**